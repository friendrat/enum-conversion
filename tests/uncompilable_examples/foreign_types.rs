/// We cannot derive a foreign trait (`TryFrom`) on
/// a foreign type (as implied by the generic parameter `U`).
///
/// This is due to error [E0210](https://doc.rust-lang.org/error_codes/E0210.html)
/// regarding orphan rules for trait implementations.
///
/// The solution is to use the `TryTo` trait provided by this crate.

use enum_conversion::prelude::*;

#[EnumConversions]
#[DeriveTryFrom]
enum Enum<'a, U> {
    F1(&'a U),
    F2(bool),
}

fn main() {

}