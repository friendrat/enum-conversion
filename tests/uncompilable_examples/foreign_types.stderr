error[E0210]: type parameter `U` must be covered by another type when it appears before the first local type (`Enum<'a, U>`)
  --> tests/uncompilable_examples/foreign_types.rs:11:1
   |
11 | #[EnumConversions]
   | ^^^^^^^^^^^^^^^^^^ type parameter `U` must be covered by another type when it appears before the first local type (`Enum<'a, U>`)
   |
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local, and no uncovered type parameters appear before that first local type
   = note: in this case, 'before' refers to the following order: `impl<..> ForeignTrait<T1, ..., Tn> for T0`, where `T0` is the first and `Tn` is the last
   = note: this error originates in the attribute macro `EnumConversions` (in Nightly builds, run with -Z macro-backtrace for more info)
